packs/BP/scripts/network.ts

```diff
@@ -89,123 +93,152 @@ export class MachineNetwork extends DestroyableObject {
-
-        let sendAmount = Math.min(
-          target.definition.maxStorage - currentAmount,
-          unsentAmount,
-        );
-
-        if (target.definition.recieveHandlerEvent) {
-          let result: number | undefined;
-
-          target.definition
-            .invokeRecieveHandler(target.block, queuedSend.type, sendAmount)
-            .then((value) => {
-              result = value;
+      // Give each machine in the low priority a split of the leftover budget (if applicable)
+      if (budget >= 0) {
+        for (let i = 0; i < machines.lowPriority.length; i++) {
+          const machine = machines.lowPriority[i];
+          const budgetAllocation = budget / (machines.lowPriority.length - i);
+          const currentStored = getMachineStorage(machine, type);
+          const machineDef = getMachineRegistration(machine);
+
+          let amountToAllocate: number = Math.min(
+            budgetAllocation,
+            machineDef.maxStorage - currentStored,
+          );
+
+          let waiting = true;
+
+          this.sendMachineAllocation(
+            machine,
+            machineDef,
+            type,
+            amountToAllocate,
+          )
+            .then((v) => {
+              amountToAllocate = v;
+              waiting = false;
             })
-            .catch((err: unknown) => {
-              logWarn(
-                `failed to call the 'recieve' handler (ID: '${target.definition.recieveHandlerEvent!}') for machine '${target.definition.id}': ${err instanceof Error || typeof err === "string" ? err.toString() : "unknown error"}. this may cause significant delays when sending storage types`,
-              );
-              result = sendAmount;
+            .catch((e: unknown) => {
+              throw e;
             });
 
-          while (result === undefined) {
-            yield;
-          }
+          while (waiting as boolean) yield;
 
-          sendAmount = result;
+          // finally give the machine its allocated share
+          budget -= amountToAllocate;
+          setMachineStorage(machine, type, currentStored + amountToAllocate);
+          yield;
         }
-
-        setMachineStorage(
-          target.block,
-          queuedSend.type,
-          currentAmount + sendAmount,
-        );
-
-        unsentAmount -= sendAmount;
-
-        yield;
       }
+    }
 
-      setMachineStorage(
-        queuedSend.block,
-        queuedSend.type,
-        Math.min(unsentAmount, queuedSend.definition.maxStorage),
-      );
+    this.sendJobRunning = false;
+  }
 
-      yield;
+  private async sendMachineAllocation(
+    machine: Block,
+    machineDef: InternalRegisteredMachine,
+    type: string,
+    amount: number,
+  ): Promise<number> {
+    // Allow the machine to change how much of its allocation it chooses to take
+    if (machineDef.recieveHandlerEvent) {
+      return await machineDef.invokeRecieveHandler(machine, type, amount);
```

Comment:
no need to `return await`, return the promise and it will automatically be flattened into this function's promise